### PR TITLE
Use SerialNumber type for signature timestamps

### LINKED ISSUE #2312: test 'Serial Number arithmetic' edge cases involving RRSIG's `{inception,expiration}`
validating a RRSIG involves comparing the current time (seconds since UNIX_EPOCH) against the {inception,expiration} fields of the RRSIG (see section 5.3.1 of RFC4035). Section 3.1.5 of RFC4034 indicates that the {inception,expiration} are not plain unsigned-integers but rather 'Serial Numbers' whose comparison logic in specified in RFC1982.

we should test what the behavior of other DNS implementations is when the value of these 32-bit {inception,expiration} fields is 1 << 31 seconds beyond UNIX_EPOCH, 1 << 32 seconds beyond UNIX_EPOCH and even beyond that.

cc @justahero

### PR DESCRIPTION
**Pull request by justahero**
This PR adds the `SerialNumber` NewType to support Serial Number Arithmetics as defined in [RFC 1982](https://www.rfc-editor.org/rfc/rfc1982). The RRSIG signature timestamps `inception` & `expiration` apply arithmetics for comparisons (see [RFC 4034, Section 3.1.5](https://datatracker.ietf.org/doc/html/rfc4034#section-3.1.5)).

Closes #2312

### CODE REVIEW

File: crates/proto/src/rr/serial_number.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+//! Number type to support Serial Number Arithmetics
+
+use std::{cmp::Ordering, ops::Add};
+
+/// Wrapper type to support Serial Number Aritthmetics as defined
```
Comment (djc) [inline crates/proto/src/rr/serial_number.rs:5]: "Aritthmetics" -> "Arithmetics".

File: crates/proto/src/rr/serial_number.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/// in RFC 1982. The signaure fields (expireation, inception) defined in RFC 4034, section 3.1.5
+/// are serial numbers.
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct SerialNumber(pub u32);
```
Comment (djc) [inline crates/proto/src/rr/serial_number.rs:9]: I think this commit should be squashed with the next one so that we can see how it is used in the same commit as the one adding the code.

I think this type could be `pub(crate)`, because it is only used in internal code so far?

File: crates/proto/src/xfer/dnssec_dns_handle.rs [hunk 3]
```diff
@@ -816,6 +816,10 @@ fn check_rrsig_validity(
     dnskey: RecordRef<'_, DNSKEY>,
     current_time: u32,
 ) -> RrsigValidity {
+    let current_time = SerialNumber(current_time);
+    let expiration = SerialNumber(rrsig.data().sig_expiration());
```
Comment (djc) [inline crates/proto/src/xfer/dnssec_dns_handle.rs:820]: Should `sig_expiration()` and `sig_inception()` return a `SerialNumber` instead? The current abstraction boundary doesn't seem very helpful to me.

File: crates/proto/src/rr/serial_number.rs (hunk 2 above)
Comment (justahero) [reply crates/proto/src/rr/serial_number.rs:9]: Commits have been squashed. The visibility needs to be `pub` when we use `SerialNumber` in other places, e.g. `sig_expiration()` & `sig_inception()`.

File: crates/proto/src/xfer/dnssec_dns_handle.rs (hunk 3 above)
Comment (justahero) [reply crates/proto/src/xfer/dnssec_dns_handle.rs]: I checked [RFC 2535](https://datatracker.ietf.org/doc/html/rfc2535) for details on the fields of the `SIG` record (as `RRSIG` wraps it). The RFC only mentions in [Appendix B, 3.](https://datatracker.ietf.org/doc/html/rfc2535#page-44) that signature expiration & inception "use serial number ring arithmetic".

Committed a change to refactor to use type `SerialNumber` for fields `expiration` & `inception`.

**Comment by justahero** [comment]
Rebased on top of #2321 for successful CI workflow.

**Review by listochkin — APPROVED** [review]

**Comment by djc** [comment]
(Would you mind also prefixing commits that only touch the `conformance` tests with "conformance: "? Would help me review more efficiently.)

**Comment by justahero** [comment]
Failing tests are due to the recent release of Rust 1.80. Waiting for #2324 to be merged.

**Comment by justahero** [comment]
The conformance specific commits have been prefixed with `conformance:`.

**Review by bluejekyll — APPROVED** [review]

**Comment by djc** [comment]
@justahero can you follow up on my question in the remaining open comment thread?

File: crates/proto/src/rr/serial_number.rs (hunk 2 above)
Comment (justahero) [reply crates/proto/src/rr/serial_number.rs]: Updated the code.

**Comment by justahero** [comment]
The visibility of the inner type has changed.

**Review by djc — APPROVED** [review]
Thanks! I had failed to notice that construction is required across crate boundaries so the privacy isn't as meaningful as I hoped. Still, let's merge this for now.
